packages/suspense/src/cache/createCache.ts

```diff
@@ -357,18 +388,14 @@ export function createCache<Params extends Array<any>, Value>(
     }
   }
 
-  function readRecordValueThrowsIfGC(record: ResolvedRecord<Value>): Value {
-    const { weakRef, value } = record.data;
-    if (weakRef != null) {
-      const retainedValue = weakRef.deref();
-      if (retainedValue == null) {
-        throw Error("Value was garbage collected");
-      } else {
-        return retainedValue;
-      }
-    } else {
-      return value!;
+  function readRecordValueThrowsIfSilentlyEvicted(
+    record: ResolvedRecord<Value>
+  ): Value {
+    const { value } = record.data;
+    if (value == null) {
+      throw Error("Value was silently evicted");
```

Comment:
This looks unsafe to me.

A cache can reasonably resolve a value to `null` or `undefined` without it meaning something was garbage collected.

I think the `CacheMap` implementation needs to be the thing that throws on read, for this to be safe. Here's my reasoning:

* Normal `Map` implementation can bypass the overhead of this check entirely.
* LRU implementation can check its internal `has()` method for this.
* `WeakRef` implementation _can_ through if `value.deref()` is null, only because no non-object values should be stored in it to begin with.

Reply:
Ah I see what you mean. 
I think for most cases, people will be using `CacheMap` implementations that are hand rolled or from an npm package like `lru-cache`. 

With that in mind, what do you think about a heuristic similar to React where `null` is considered a valid resolved value while `undefined` is considered a miss? We can enforce that with types, and make this check stricter by checking only for `undefined`.

Reply:
> With that in mind, what do you think about a heuristic similar to React where `null` is considered a valid resolved value while `undefined` is considered a miss?

I'm a bit uncomfortable with that as a built-in limitation of the library. I generally prefer using `null` over `undefined` for "non values" _but_ it's subjective and I know of at least one use case (at @replay– where this library is being used) where `undefined` _is a valid value_.

> We can enforce that with types, and make this check stricter by checking only for `undefined`.

Not everyone uses TypeScript, so I don't think we can/should rely on static type checking to enforce soundness either.